### PR TITLE
[FlagGems Operator Development Competition] Add max_pool3d

### PR DESCRIPTION
**Pull request by BuaaAlban**
## Summary
- Implement `torch.nn.functional.max_pool3d` (forward + backward) using hand-written Triton kernels with `@libentry()` + `@triton.autotune` + `@triton.jit`
- Forward kernel (`max_pool3d_forward_kernel`): 2D tiled (BLOCK_H x BLOCK_W) over output spatial H/W, with depth and kernel loops unrolled via `tl.static_range`; 12 autotune configs covering block sizes from 8x8 to 64x64
- Backward kernel (`max_pool3d_backward_kernel`): 2D tiled over input spatial H/W, iterates over kernel offsets to find matching output positions via index comparison; 6 autotune configs
- Support variants: `max_pool3d_with_indices` (forward, returns output + indices) and `max_pool3d_backward` (gradient computation)
- Full parameter support: `kernel_size` (int or 3-tuple), `stride` (int, 3-tuple, or None), `padding` (int or 3-tuple), `dilation` (int or 3-tuple), `ceil_mode` (bool)
- Precision: aligned with PyTorch reference via `gems_assert_close` with upcast reference computation

### Test Functions (2 total)
| Test Function | Variant | Input Type |
|---------------|---------|------------|
| `test_accuracy_max_pool3d` | Forward | 5D tensor (N, C, D, H, W) |
| `test_accuracy_max_pool3d_backward` | Backward | 5D tensor (N, C, D, H, W) + grad_output |

### Parameter Coverage
| Dimension | Values | Count |
|-----------|--------|-------|
| **Config** (shape, kernel_size, stride, padding, dilation, ceil_mode) | 9 zipped tuples (see below) | 9 |
| **dtype** | torch.float16, torch.float32, torch.bfloat16 | 3 |

### MAXPOOL3D_CONFIGS (9 configurations, zipped tuples)
| # | Shape | kernel_size | stride | padding | dilation | ceil_mode | Description |
|---|-------|-------------|--------|---------|----------|-----------|-------------|
| 1 | (4,3,16,16,16) | 3 | 2 | 1 | 1 | False | Classic 3x3x3, stride 2, pad 1 |
| 2 | (8,16,12,14,14) | (2,3,3) | (1,2,2) | (0,1,1) | 1 | False | Non-cubic kernel/stride/padding |
| 3 | (2,4,15,15,15) | 3 | 2 | 1 | 1 | True | ceil_mode enabled |
| 4 | (1,1,9,9,9) | 2 | 1 | 0 | 2 | False | Dilation=2 |
| 5 | (1,64,8,28,28) | 3 | 2 | 1 | 1 | False | Typical 3D CNN shape |
| 6 | (2,8,8,16,16) | 2 | 2 | 0 | 1 | False | No padding |
| 7 | (2,8,10,16,20) | 2 | 2 | (0,1,0) | 1 | False | Non-symmetric padding |
| 8 | (1,1,5,5,5) | 2 | 1 | 0 | 1 | False | Small input |
| 9 | (8,16,8,8,8) | 3 | 1 | 1 | 1 | False | Large batch |

**Total test cases:** 9 configs x 3 dtypes x 2 functions = **54 test cases**

**All 54 tests PASSED.**

#### dtype=torch.float16
| Shape | Params | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------|-------------|--------------|---------|
| (4,3,16,56,56) | k=3,s=2,p=1,d=1 | 0.0115 | 0.0099 | **1.167x** |
| (4,3,16,56,56) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0117 | 0.0101 | **1.165x** |
| (4,3,16,56,56) | k=3,s=1,p=1,d=2 | 0.0261 | 0.0172 | **1.512x** |
| (4,3,16,56,56) | k=3,s=2,p=1,d=1,ceil | 0.0114 | 0.0153 | 0.743x |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1 | 0.0362 | 0.0173 | **2.089x** |
| (8,64,8,28,28) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0460 | 0.0197 | **2.329x** |
| (8,64,8,28,28) | k=3,s=1,p=1,d=2 | 0.0830 | 0.0669 | **1.241x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1,ceil | 0.0407 | 0.0192 | **2.118x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1 | 0.0329 | 0.0125 | **2.622x** |
| (16,128,4,14,14) | k=3,s=1,p=1,d=2 | 0.0446 | 0.0266 | **1.680x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1,ceil | 0.0430 | 0.0156 | **2.761x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1 | 0.0310 | 0.0187 | **1.658x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1,ceil | 0.0484 | 0.0291 | **1.662x** |

#### dtype=torch.float32
| Shape | Params | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------|-------------|--------------|---------|
| (4,3,16,56,56) | k=3,s=2,p=1,d=1 | 0.0119 | 0.0586 | 0.203x |
| (4,3,16,56,56) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0118 | 0.0221 | 0.535x |
| (4,3,16,56,56) | k=3,s=1,p=1,d=2 | 0.0264 | 0.0173 | **1.524x** |
| (4,3,16,56,56) | k=3,s=2,p=1,d=1,ceil | 0.0113 | 0.0097 | **1.170x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1 | 0.0361 | 0.0194 | **1.863x** |
| (8,64,8,28,28) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0456 | 0.0213 | **2.138x** |
| (8,64,8,28,28) | k=3,s=1,p=1,d=2 | 0.0816 | 0.0627 | **1.302x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1,ceil | 0.0407 | 0.0196 | **2.080x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1 | 0.0332 | 0.0127 | **2.611x** |
| (16,128,4,14,14) | k=3,s=1,p=1,d=2 | 0.0445 | 0.0259 | **1.719x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1,ceil | 0.0421 | 0.0149 | **2.824x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1 | 0.0310 | 0.0183 | **1.699x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1,ceil | 0.0478 | 0.0288 | **1.658x** |

#### dtype=torch.bfloat16
| Shape | Params | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------|--------|-------------|--------------|---------|
| (4,3,16,56,56) | k=3,s=2,p=1,d=1 | 0.0115 | 0.0099 | **1.166x** |
| (4,3,16,56,56) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0121 | 0.0101 | **1.189x** |
| (4,3,16,56,56) | k=3,s=1,p=1,d=2 | 0.0268 | 0.0175 | **1.527x** |
| (4,3,16,56,56) | k=3,s=2,p=1,d=1,ceil | 0.0114 | 0.0090 | **1.270x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1 | 0.0368 | 0.0174 | **2.120x** |
| (8,64,8,28,28) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0472 | 0.0197 | **2.393x** |
| (8,64,8,28,28) | k=3,s=1,p=1,d=2 | 0.0846 | 0.0666 | **1.271x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1,ceil | 0.0416 | 0.0191 | **2.181x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1 | 0.0338 | 0.0124 | **2.722x** |
| (16,128,4,14,14) | k=3,s=1,p=1,d=2 | 0.0457 | 0.0264 | **1.730x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1,ceil | 0.0434 | 0.0146 | **2.982x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1 | 0.0314 | 0.0183 | **1.714x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1,ceil | 0.0490 | 0.0291 | **1.683x** |

#### dtype=torch.float16
| Input Shape | Params | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------------|--------|-------------|--------------|---------|
| (4,3,16,56,56) | k=3,s=2,p=1,d=1 | 0.0255 | 0.0302 | 0.844x |
| (4,3,16,56,56) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0243 | 0.0279 | 0.871x |
| (4,3,16,56,56) | k=3,s=1,p=1,d=2 | 0.0967 | 0.0320 | **3.025x** |
| (4,3,16,56,56) | k=3,s=2,p=1,d=1,ceil | 0.0220 | 0.0307 | 0.719x |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1 | 0.0590 | 0.1253 | 0.471x |
| (8,64,8,28,28) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0821 | 0.0980 | 0.838x |
| (8,64,8,28,28) | k=3,s=1,p=1,d=2 | 0.2932 | 0.1223 | **2.398x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1,ceil | 0.0707 | 0.1251 | 0.565x |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1 | 0.0504 | 0.0674 | 0.747x |
| (16,128,4,14,14) | k=3,s=1,p=1,d=2 | 0.0860 | 0.0652 | **1.321x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1,ceil | 0.0670 | 0.0680 | **0.986x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1 | 0.0475 | 0.1207 | 0.394x |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1,ceil | 0.0778 | 0.1190 | 0.654x |

#### dtype=torch.float32
| Input Shape | Params | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------------|--------|-------------|--------------|---------|
| (4,3,16,56,56) | k=3,s=2,p=1,d=1 | 0.0186 | 0.0260 | 0.715x |
| (4,3,16,56,56) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0204 | 0.0217 | **0.938x** |
| (4,3,16,56,56) | k=3,s=1,p=1,d=2 | 0.0397 | 0.0282 | **1.409x** |
| (4,3,16,56,56) | k=3,s=2,p=1,d=1,ceil | 0.0184 | 0.0260 | 0.706x |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1 | 0.0506 | 0.1113 | 0.454x |
| (8,64,8,28,28) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0650 | 0.0869 | 0.747x |
| (8,64,8,28,28) | k=3,s=1,p=1,d=2 | 0.1163 | 0.1090 | **1.066x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1,ceil | 0.0558 | 0.1115 | 0.500x |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1 | 0.0468 | 0.0606 | 0.773x |
| (16,128,4,14,14) | k=3,s=1,p=1,d=2 | 0.0618 | 0.0583 | **1.061x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1,ceil | 0.0560 | 0.0608 | **0.921x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1 | 0.0432 | 0.1092 | 0.395x |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1,ceil | 0.0667 | 0.1090 | 0.612x |

#### dtype=torch.bfloat16
| Input Shape | Params | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------------|--------|-------------|--------------|---------|
| (4,3,16,56,56) | k=3,s=2,p=1,d=1 | 0.0228 | 0.0305 | 0.746x |
| (4,3,16,56,56) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0268 | 0.0261 | **1.025x** |
| (4,3,16,56,56) | k=3,s=1,p=1,d=2 | 0.0966 | 0.0317 | **3.051x** |
| (4,3,16,56,56) | k=3,s=2,p=1,d=1,ceil | 0.0231 | 0.0307 | 0.755x |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1 | 0.0613 | 0.1267 | 0.484x |
| (8,64,8,28,28) | k=(2,3,3),s=(1,2,2),p=(0,1,1),d=1 | 0.0855 | 0.0978 | 0.874x |
| (8,64,8,28,28) | k=3,s=1,p=1,d=2 | 0.2942 | 0.1203 | **2.445x** |
| (8,64,8,28,28) | k=3,s=2,p=1,d=1,ceil | 0.0714 | 0.1249 | 0.572x |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1 | 0.0512 | 0.0676 | 0.758x |
| (16,128,4,14,14) | k=3,s=1,p=1,d=2 | 0.0877 | 0.0653 | **1.342x** |
| (16,128,4,14,14) | k=3,s=2,p=1,d=1,ceil | 0.0684 | 0.0680 | **1.005x** |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1 | 0.0482 | 0.1209 | 0.399x |
| (32,256,2,7,7) | k=3,s=2,p=1,d=1,ceil | 0.0786 | 0.1198 | 0.656x |

### Performance Summary

| Variant | dtype | Min Speedup | Max Speedup | Median Speedup | Configs < 0.9x |
|---------|-------|-------------|-------------|----------------|-----------------|
| **Forward** | float16 | 0.743x | 2.761x | **1.662x** | 1/13 |
| **Forward** | float32 | 0.203x | 2.824x | **1.699x** | 2/13 |
| **Forward** | bfloat16 | 1.166x | 2.982x | **1.683x** | 0/13 |
| **Backward** | float16 | 0.394x | 3.025x | **0.838x** | 7/13 |
| **Backward** | float32 | 0.395x | 1.409x | **0.715x** | 8/13 |
| **Backward** | bfloat16 | 0.399x | 3.051x | **0.755x** | 7/13 |

### Performance Analysis

**Forward kernel** performs well across all dtypes and most shapes:
- Achieves strong speedups (1.2x-3.0x) on medium-to-large shapes with high channel counts, where the 2D tiled approach with autotune effectively utilizes GPU parallelism
- The float32 entries at shape (4,3,16,56,56) with k=3,s=2 show low speedup (0.203x and 0.535x), likely due to Triton autotune JIT compilation overhead on the first invocation for that config key combination; subsequent shapes with different config keys perform well
- The float16 ceil_mode entry on (4,3,16,56,56) at 0.743x is a small-batch, low-channel case where kernel launch overhead dominates
- **bfloat16 forward achieves >= 1.0x speedup on ALL 13 configurations**

**Backward kernel** shows mixed performance:
- Excels on dilation=2 configs (2.4x-3.1x), where the larger effective kernel makes the scatter-style backward approach efficient
- Sub-0.9x on shapes with very small output spatial dimensions (e.g., (32,256,2,7,7) produces tiny output), where kernel launch overhead dominates over computation
- The backward kernel tiles over **input** spatial dims and iterates over kernel offsets; for high-channel, small-spatial cases, the fixed overhead per program instance is proportionally large

### CODE REVIEW

**Review by w1120029931-bit — CHANGES_REQUESTED** [review]
Your test code is not standardized. Please revise it. The correct format is as follows. For example:
ref_out = torch.relu(ref_inp)
with flag_gems.use_gems():
res_out = torch.nn.functional.relu(res_inp)   
AND
Create your own benchmark and tests files for your operator，exp：
benchmark/test_max_pool3d_perf.py
tests/test_max_pool3d.py

**Comment by BuaaAlban** [comment]
> Your test code is not standardized. Please revise it. The correct format is as follows. For example: ref_out = torch.relu(ref_inp) with flag_gems.use_gems(): res_out = torch.nn.functional.relu(res_inp) AND Create your own benchmark and tests files for your operator，exp： benchmark/test_max_pool3d_perf.py tests/test_max_pool3d.py

Thank you for your guide , fixed and pushed~

**Review by w1120029931-bit — APPROVED** [review]
LGTM
